gr-digital/lib/decision_feedback_equalizer_impl.cc

```diff
@@ -0,0 +1,267 @@
+                                  const std::string& training_start_tag)
+{
+    return gnuradio::get_initial_sptr(
+        new decision_feedback_equalizer_impl(num_taps_forward,
+                                             num_taps_feedback,
+                                             sps,
+                                             alg,
+                                             adapt_after_training,
+                                             training_sequence,
+                                             training_start_tag));
+}
+
+/*
+ * The private constructor
+ */
+decision_feedback_equalizer_impl::decision_feedback_equalizer_impl(
+    unsigned num_taps_forward,
+    unsigned num_taps_feedback,
+    unsigned sps,
+    adaptive_algorithm_sptr alg,
+    bool adapt_after_training,
+    std::vector<gr_complex> training_sequence,
+    const std::string& training_start_tag)
+    : gr::sync_decimator("decision_feedback_equalizer",
+                         io_signature::make(1, 1, sizeof(gr_complex)),
+                         io_signature::makev(1,
+                                             3,
+                                             std::vector<int>{ sizeof(gr_complex),
+                                                               (int)((num_taps_forward +
+                                                                      num_taps_feedback) *
+                                                                     sizeof(gr_complex)),
+                                                               sizeof(unsigned short) }),
+                         sps),
+      filter::kernel::fir_filter_ccc(
+          sps,
+          vector<gr_complex>(num_taps_forward + num_taps_feedback, gr_complex(0, 0))),
+      d_num_taps_fwd(num_taps_forward),
+      d_num_taps_rev(num_taps_feedback),
+      d_sps(sps),
+      d_alg(alg),
+      d_adapt_after_training(adapt_after_training),
+      d_training_sequence(training_sequence),
+      d_training_start_tag(training_start_tag),
+      d_num_taps(num_taps_forward + num_taps_feedback),
+      d_decision_history(num_taps_feedback),
+      d_new_taps(num_taps_forward + num_taps_feedback),
+      d_updated(false),
+      d_training_sample(0),
+      d_filt_buf(num_taps_forward + num_taps_feedback)
+{
+    // not reversed at this point
+    d_new_taps[0] = gr_complex(1.0, 0.0); // help things along with initialized taps
+
+    if (training_start_tag == "" || training_sequence.empty()) {
+        d_training_state = equalizer_state_t::DD;
+    } else {
+        d_training_state = equalizer_state_t::IDLE;
+    }
+
+    set_history(d_num_taps_fwd);
+
+    alg->initialize_taps(d_new_taps);
+    // NOTE: the filter kernel reverses the taps internally
+    filter::kernel::fir_filter_ccc::set_taps(d_new_taps);
+}
+
+void decision_feedback_equalizer_impl::update_decision_history(gr_complex decision)
+{
+    std::rotate(d_decision_history.begin(),
```

Comment:
This is called a lot, in a loop from `equalize()`, and in turn in from `work()`. `d_decision_history` is a vector. That's a lot of data copying, isn't it?

Did you consider a `std::queue` for this instead? Just `d_decision_history.pop(); d_decision_history.push(…);`?

Maybe I'm missing something.

Reply:
d_decision_history gets copied into the filter buffer (along with the incoming samples) which needs direct element access.  I thought queues didn't have iterators so there was no way to copy a queue to an array.   I can see what I've implemented there is a queue, but don't see an easy way around it.

Reply:
Doh, of course. Just switch to `std::deque`, which is the default container class for `std::queue` anyway, but has iterators. Just use `push_back` instead of `push` in my example.